WIP: v2.1 (prev: Architecture Rework)

NovusTheory opened the pull request:
Edit: This PR used to be just about an architecture rework but now it's v2.1

This is a complete internal architecture and codebase rework which helps improve the readability and functionality of the app from the initial v2 version that had more of a MVP codebase. As more and more gets added it becomes more obvious that this change needs to happen sooner than later to ensure the codebase doesn't become a mess.

### If you are a developer contributing to YTMDesktop this PR will affect any current work as it is a very heavy codebase change among implementing tons of other features

Notable changes
- Singleton service style architecture
   - Watchdog
       - Crash and log handling
   - Auto updater
       - Also changes the app to open an updater window on startup just for update checking before launching the main program
   - Config store
       - Everything inside config.json
   - State manager
       - Handles the current page url and window bounds to flush them to disk more frequently
   - Integration manager
       - Introduces lifecycle runs to integrations and a base class for more ease of programming them
   - YTM view manager
       - Handles everything related to showing YouTube Music
       - Other systems to show YTM regardless if the app can hook it or not
           - A warning dialog is shown when hook errors happen to alert the user functionality is degraded but not limiting usage of YTM entirely
   - Window manager
       - Handles all windows and views in the app and can automatically recreate them if they're closed abnormally e.g. crashed
       - Manages automatic sizing for views attached to windows
   - Protocol manager
       - Manages the ytmd:// protocol
       - Manages the internal ytmd-app:// protocol which is used to serve the apps files and [follows Electrons best practices for security](https://www.electronjs.org/docs/latest/tutorial/security#18-avoid-usage-of-the-file-protocol-and-prefer-usage-of-custom-protocols)

Feature additions
- Timed Lyrics
- shuffleEnabled state in the companion API
- Queue modification commands in the companion API
- Read-only IPC server for local apps to read the state of YTM
- Window title now displays `Video Title - Author` and more consistently
- Updated discord rich presence features
- Enhanced media service via Xosms
- Miniplayer
- mDNS broadcasting (_ytmdesktop._tcp) to find active instances of a companion server on the network. Useful for mobile app developers or anything off the same system as YTMD

NovusTheory committed:
✨ Enhanced media service via xosms

NovusTheory committed:
⬆️ Update xosms to v0.4.0

NovusTheory committed:
💚 Add libdbus to CI build

NovusTheory committed:
💚 Wrong if statement for CI

NovusTheory committed:
🩹 Remove indented from enhanced media session setting

NovusTheory committed:
🩹 Add restart-required to enhanced media session

NovusTheory committed:
🩹 Update media service less often

NovusTheory committed:
⬆️ Upgrade xosms

NovusTheory committed:
🐛 Fix event handlers from media service

NovusTheory committed:
✨ Upgrade xosms with seek support

NovusTheory committed:
🐛 Fix seeking from media service

NovusTheory committed:
🩹 Add beta tag and disable on macos

NovusTheory committed:
🐛 Invert macos check

NovusTheory committed:
🩹 Restore restart requirement for enhanced media service

NovusTheory committed:
🚸 Update disabled message for enhanced media service

NovusTheory committed:
⬆️ Upgrade xosms for bug fixes

NovusTheory committed:
⬆️ Upgrade xosms with more bug fixes

NovusTheory committed:
✨ Initial timed lyrics feature

NovusTheory committed:
🏗️ Initial rework of app architecture

This improves various aspects of the app internally for better maintainability

NovusTheory committed:
🩹 Add prettyPrint false to timed lyrics request

NovusTheory committed:
👔 Change timed lyrics logic to use innertube network manager

NovusTheory committed:
✨ Add scroll to current and autoscroll disabling systems on user scroll

NovusTheory committed:
🩹 Replace YTM lyrics when swapping between video and song

Music videos don't have timed lyrics and will be out of sync if we show timed lyrics

NovusTheory committed:
💬 "Scroll to Current" -> "Sync to video time"

NovusTheory committed:
🧱 Update for vite changes

NovusTheory committed:
🧱 Rebase with developement branch

NovusTheory committed:
🐛 Fix loadURL vs loadFile for packaged app

NovusTheory committed:
✨ Implement hook flags

More detailed system to show where the failure is when hooking YTM

NovusTheory committed:
🐛 Wrongful double function call for ExtrasSetup

NovusTheory committed:
💡 Update comment for reconcileConfig in ConfigStore

NovusTheory committed:
🐛 Fix pause timeout for Discord presence

NovusTheory committed:
🩹 Disable autoRecreate for windows

This is being changed as the behavior isn't really expected and tends to lead to unexpected behavior more than it should compared to if a view was auto recreated

NovusTheory committed:
🏗️ Swap to a ServiceHost based system

Currently excludes playerStateStore for now

Akiisqt commented on src/main/integrations/discord-presence/index.ts:
```diff
@@ -49,11 +49,11 @@ function stringLimit(str: string, limit: number, minimum: number) {
 
   private discordClient: DiscordClient = null;
-  private enabled = false;
   private ready = false;
```
isReady? To follow the change to isEnabled.

Akiisqt commented on src/main/integrations/discord-presence/index.ts:
```diff
@@ -123,7 +123,7 @@ export default class DiscordPresence implements IIntegration {
       hasFullMetadata &&
       (oldState !== this.videoState || oldId !== this.videoDetails.id || Math.abs(this.progress - oldProgress) > 1 || oldProgress > this.progress)
```
Have you looked into just using the metadata state loaded or whatever for this? And if so how did it go

Alipoodle replied:
Considering this isn't changed in this PR this is irrelevant to this.

Akiisqt commented on src/main/integrations/integration.ts:
```diff
@@ -1,9 +1,85 @@
+  /**
+   * Disables the integration
+   */
+  public async disable(): Promise<NoOverride> {
```
onDisable to match other stuff

A-Matt replied:
(See enable's for reason)

Alipoodle replied:
See enable

Akiisqt commented on src/main/integrations/integration.ts:
```diff
@@ -1,9 +1,85 @@
+  /**
+   * Enables the integration
+   */
+  public enable(): NoOverride {
```
onEnable or setup to match other changes to be more consistent

A-Matt replied:
```ts
// Enforces TypeScript to not allow overriding a method (MUST NOT BE EXPORTED)
declare const _never: unique symbol;
type NoOverride = { [_never]: typeof _never };
```
This as well as the `disable` function are designed as more "internal" APIs to toggle the internal flag of the `isEnabled` variable and then run the `onEnabled` and `onDisabled` functions afterwards.

This basically eliminates the  problem of having to handle a bunch of the additional stuff that is already managed within each individual plugin, and instead centralise it.

Alipoodle replied:
These are the generic functions to enable / disable the integration.

https://github.com/ytmdesktop/ytmdesktop/pull/1469/files#diff-9ecdb46030bbd87a6b123161a8f0454c994b4987119ec355e8da7a339384c8a1R78 is designed to be the overwritten function in an integration to run after the enable function has been fired.

Running `<Integration>.enable()` is fine

Akiisqt commented on src/main/integrations/integration.ts:
```diff
@@ -1,9 +1,85 @@
+   * This is always run after the app has emitted ready
+   */
+  public abstract onEnabled(): void;
+  /**
```
I see this now, why is there two enabled functions?

Alipoodle replied:
On enabled = Integration overwrite and will be to enable functionality of the integration
Enable = Turn the integration on and flip the flag, then fire the on enabled function

Akiisqt commented on src/main/services/autoupdater/index.ts:
```diff
@@ -0,0 +1,182 @@
+    return false;
+  }
+
+  public override onPreInitialized() {}
```
What would this be used for

Akiisqt reviewed (commented):
Left a review on some stuff otherwise good job so far

NovusTheory committed:
🐛 Fix timed lyrics by making it an anonymous request.

For context Google has add more request verification and we can't route this request through YTMs network manager anymore utilizing the context hack

NovusTheory committed:
🚸 Add note that timed lyrics is being provided by YTMDesktop

NovusTheory committed:
🐛 Fix updater window logo reference

NovusTheory committed:
💡 Remove TODO comment for setupCompletions flag

NovusTheory committed:
✨ Add ProtocolManager

This manages the ytmd:// protocol on the system and the internal ytmd-app:// protocol that is used to serve the local apps files

NovusTheory committed:
🔨 Turn off file protocol extra privileges

NovusTheory committed:
✨ Add queueAdd, queueRemove, and queueMove to the API

NovusTheory committed:
🩹 Change NODE_ENV checks to isPackaged

NovusTheory committed:
🐛 Fix history button display

NovusTheory committed:
🏗️ Ensure preload doesn't run and ensure YTMView is shown when not on YTM

NovusTheory committed:
🏗️ Add window open handler for windows & impl for settings window

NovusTheory committed:
🐛 Fix companion auth window being blank when packaged

NovusTheory committed:
✨ Add new rich presence features to Discord

NovusTheory committed:
✨ Add new and enhanced context menu within YTM

NovusTheory committed:
⚰️ Clean up some dead code in comments

NovusTheory committed:
💡 Update comment for _getElectronWindow

NovusTheory committed:
🚸 Re-add slow loading message and ensure unresponsive text works in titlebar

NovusTheory committed:
🐛 Fix #1563

NovusTheory committed:
🏗️ ⬆️ Upgrade dependencies and run main app as ESM

NovusTheory committed:
✨ Add ipc server to companion server integration

This implements a read-only ipc server which slightly mimics what our socket.io server does but strictly adheres to a single picked default version and cannot be changed

NovusTheory committed:
🚸 Default enable browser communication for companion server

NovusTheory committed:
🚸 Implement window title to match page title

NovusTheory committed:
✨ Add more shortcuts

volumeMute, toggleShuffle, toggleRepeat (cycle)

NovusTheory committed:
✨ Add shuffleEnabled state to companion api

Fixes #1542

NovusTheory committed:
♻️ Rework window title to utilize state store

NovusTheory committed:
🚨 Disable lint problems for ytmview scripts

NovusTheory committed:
💬 Update note about timed lyrics

NovusTheory committed:
🐛 Rebase error fixes

NovusTheory committed:
🐛 Fix queueAdd not allowing to add to end of queue

NovusTheory committed:
⬆️ 🚨 🐛 Rebase fixes, patch for ESM to Xosms, and lint upgrade

This includes changes from another local branch that had work done

NovusTheory committed:
🚨 Lint run with updated eslint

NovusTheory committed:
🐛 Set user gesture for remote control to allow playing on startup from media service

NovusTheory committed:
🐛 Deactivate media player on construction & wait for hasFullMetadata

NovusTheory committed:
🔧 Add auto unpack natives to forge config

NovusTheory committed:
✨ Add mdns service for companion server discovery

NovusTheory committed:
🏗️ Split player state store types out into shared

NovusTheory committed:
✨ Initial creation of miniplayer

NovusTheory committed:
💄 Improve UI of miniplayer

NovusTheory committed:
🐛 Hydrate state on initial application startup

NovusTheory committed:
💄 Inline author with title

NovusTheory committed:
🚸 Remember miniplayer window bounds

NovusTheory committed:
🐛 Fix missing miniplayer window in packaged app

NovusTheory committed:
🚸 Add clipboard copying and bug report message to hook error

NovusTheory committed:
🐛 Fix application state hydration on launch where it can sometimes fail

Currently an unreliable method but sometimes it works

NovusTheory committed:
✨ Support version selection for IPC

NovusTheory committed:
🏗️ Remove hello packet from IPC server

Clients are expected to query /metadata for versions instead

NovusTheory committed:
🔇 Remove some debug logging

NovusTheory committed:
🚸 Lower font size of miniplayer titlebar icons & remove unnecessary reconcile

NovusTheory committed:
🏗️ 🐛 Make player state store a service & fix EventEmitterService emits

NovusTheory committed:
🏗️ Move YTMView scripts to a MessagePort based API system

NovusTheory committed:
🐛 Remove waiting invoke when it's called

NovusTheory committed:
💄 Change miniplayer to a frosted glass look

NovusTheory committed:
🚸 Add prevent-unload prompt

NovusTheory committed:
🚸 Add refresh YTM button if YTM fails to load

NovusTheory committed:
🚸 Check for YTM domain before saving url

Saving the url if the user is currently in a login flow or something else isn't ideal. This prevents less confusion.

NovusTheory committed:
💄 Add repeat mode button to miniplayer, fix drag handle position, and topbar blur

NovusTheory committed:
🐛 Correct various integrations and services to the RemoteControl PAPI

RatioVolume scripts will use an old method for now

NovusTheory committed:
🚸 Update drag regions for miniplayer

NovusTheory committed:
🐛 Add executeScript ipc for ratio volume integration

NovusTheory committed:
✨ Add more player controls to the miniplayer

NovusTheory committed:
🚸 Update always on top to be pop-up-menu

NovusTheory committed:
🐛 Fix lastVideoId/playlistId not being saved

NovusTheory committed:
👔 Update autoupdater logic

NovusTheory committed:
✨ Add changelog window

NovusTheory committed:
🐛 Set volume on miniplayer if state changes volume

NovusTheory committed:
🐛 Generic changes and fixes

- ProtectedAPI port closing
- Autoplay policy changes
- Volume ratio fixes if YTMView is recreated

NovusTheory committed:
✨ Add flagmanager & integration kill switches

NovusTheory committed:
👔 Add flag change broadcast & reorder service load

NovusTheory committed:
👔 Add migration for appVersion in metadata of config

NovusTheory committed:
🚸 Add aspect ratio enforcement to miniplayer

NovusTheory committed:
🐛 Fix polymer hooking

NovusTheory committed:
⬆️ Bump electron to latest

NovusTheory committed:
🐛 Change version to 2.1.0-dev
